강원대 BE 최현준 - Step3 데이터베이스 적용

Hyeonjun0527 opened the pull request:
### 변경사항

1. application.properties
2. schema.sql
3. ProductRepositoryImpl의 CRUD 기능

### 중점적으로 리뷰받고 싶은 곳
ProductRepositoryImpl의 CRUD 로직에 문제는 없는지 리뷰받고 싶습니다.

서비스 계층에 트랜잭션을 두면 하나의 비즈니스 작업이 전부 함께 커밋·롤백돼서 흐름을 이해하기 쉽고 데이터 일관성이 잘 지켜질 거 같습니다. 그런데 다른 객체가 제 ProductReopsitory를 사용하다가 트랜잭션실수를 저지를 수 있을 것 같습니다. 개발을 진행하다가 매번 다른 도메인에게 "이 서비스 만들어주세요." 이렇게 요청하다보니 개발 진행속도가 매우 느려진 경험을 한 적이 있습니다. 그렇게 되니 다른 도메인의 레포지토리를 많이 쓰게 되는 것 같기도 하고요. 이에 대한 멘토님의 생각이 궁금합니다.

보통 도메인-주도 설계 관점에서는 한 도메인의 Repository 는 그 도메인 내부 구현 세부사항으로 간주하고, 외부에서는 Service(또는 애그리것 루트가 제공하는 퍼블릭 메서드)를 통해 접근하라고 권장한다고 하는데 그때 생기는 개발 속도 지연이 궁금합니다.

Hyeonjun0527 committed:
chore: h2 db 환경설정
build.gradle에 의존성 추가
application.properties에 h2 db 설정 추가
h2 콘솔 활성화 및 jdbc url, username,password

Hyeonjun0527 committed:
chore: schema.sql 생성
Product DDL 작성

Hyeonjun0527 committed:
feat : README.md 작성

Hyeonjun0527 committed:
fix(schema.sql): 스키마 에러 해결

Hyeonjun0527 committed:
refactor(product): Read 기능 H2 조회 개선 및 JDBC 활용으로 전환
- JdbcClient와 RowMapper를 활용한 SQL 쿼리 처리 방식 도입

Hyeonjun0527 committed:
refactor(ProductRepositoryImpl): Create와 Update 기능 마이그레이션 및 JDBC 활용으로 전환

Hyeonjun0527 committed:
refactor(ProductRepositoryImpl): Delete기능 구현 Delete문과 Exist문 작성

woowahan-justin commented on README.md:
```diff
@@ -26,4 +26,26 @@ DELETE /api/products/{productId}
\ No newline at end of file
+4. 관리자 상품 삭제 기능
+
+## 3단계
```
요구사항 정리 👍

woowahan-justin commented on src/main/java/gift/domain/product/repository/ProductRepositoryImpl.java:
```diff
@@ -1,62 +1,129 @@
 
-        if (start > productList.size()) {
-            return new PageImpl<>(Collections.emptyList(), pageable, productList.size());
+        int totalRow = jdbcClient.sql("""
+                        SELECT COUNT(*)
+                        FROM Product;
+                """).query(Integer.class).single();
```
Count 를 가져오는 책임은 자주 사용될 수 있으니 별도의 메서드로 분리해도 좋을 것 같아요 :)

woowahan-justin commented on src/main/java/gift/domain/product/repository/ProductRepositoryImpl.java:
```diff
@@ -1,62 +1,129 @@
-        return products.get(id);
+    public Optional<Product> findById(Long id) {
+
+        return id == null ? null : jdbcClient.sql(
```
null 을 피하기 위해 Optional 을 사용하는데요
Optional 을 잘 활용하기 위해선 어떻게 처리하는 것이 좋을까요?

woowahan-justin commented on src/main/java/gift/domain/product/repository/ProductRepositoryImpl.java:
```diff
@@ -1,62 +1,129 @@
+
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        Long id = product.getId();
+        if (id == null) {
+            jdbcClient.sql(
+                            """
+                                    INSERT INTO PRODUCT (name, price, image_url)
+                                    VALUES (:name, :price, :imageUrl)
+                                    """
+                    ).param("name", product.getName())
+                    .param("price", product.getPrice())
+                    .param("imageUrl", product.getImageUrl())
+                    .update(keyHolder);
+
+            id = Objects.requireNonNull(keyHolder.getKey()).longValue();
+        } else {
+            jdbcClient.sql(
+                            """
+                                         UPDATE PRODUCT
+                                         SET name = :name,
+                                             price = :price,
+                                             image_url = :imageUrl
+                                         WHERE id = :id
+                                    """
+                    ).param("id", id)
+                    .param("name", product.getName())
+                    .param("price", product.getPrice())
+                    .param("imageUrl", product.getImageUrl())
+                    .update();
+
```
else 를 사용하지 않고 처리할 수 있을 것 같아요 :)
코드의 가시성을 위해 메서드로 분리해보면 어떨까요?

woowahan-justin reviewed (changes requested):
안녕하세요 현준님 :)
3단계 미션 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~
확인하고 다시 요청해 주세요 🙇

Hyeonjun0527 committed:
refactor(product-repository): 코드 리팩토링

- getProductTotalRow 메서드를 도입
- insertProduct와 updateProduct 메서드로 로직 분리
- findById 메서드에서 조회 결과가 없는 경우 Optional.empty() 반환
 (NullPointerException 방지)

Hyeonjun0527 committed:
refactor(all): 개행문자 삽입 및 줄바꿈 정리

Hyeonjun0527 committed:
refactor(exception-handling): 에러 처리 방식 개선

- ErrorField 제거 및 ErrorCode에 HTTP 상태 코드 필드 추가
- 글로벌 예외 처리기에서 ErrorCode를 활용
- 메서드 인자 유효성 검사 예외 핸들링 로직 추가
- 수정된 예외 처리에 대한 테스트 코드 업데이트

Hyeonjun0527 committed:
refactor(exception): 예외 처리 로직 리팩터링

- ValidationExceptionHandler를 새로 추가하여 검증 관련 예외를 독립적으로 처리하도록 분리
- GlobalExceptionHandler를 단순화 및 역할 분리
  - NullPointerException, NumberFormatException 등 주요 예외 처리 유지
- ErrorResponse 사용 대신 ProblemDetail 사용(Spring 6+의 공식 표준은 ProblemDetail)
- ProductResponse 생성자를 private로 변경

liquidjoo commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -1,106 +1,53 @@
 @RestControllerAdvice
-@Order
-public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+@Order(1)
```
Order 를 사용 시 주의해야할 점들이 있는데요! 후에 충돌이 나지 않도록 주의해보면 좋을 것 같습니다

liquidjoo commented on src/main/java/gift/domain/product/repository/ProductRepositoryImpl.java:
```diff
@@ -1,62 +1,134 @@
+        this.jdbcClient = jdbcClient;
+    }
+
+    private static final RowMapper<Product> PRODUCT_MAPPER = (rs, rowNum) -> new Product(
+            rs.getLong("id"),
+            rs.getString("name"),
+            rs.getInt("price"),
+            rs.getString("image_url")
+    );
```
별도의 역할로 분리해서 만들어보아도 좋아요
어떤 장점들이 있을까요?

liquidjoo commented on src/main/java/gift/common/exception/ErrorCode.java:
```diff
@@ -1,29 +1,43 @@
-    MALFORMED_JSON("잘못된 JSON 형식입니다."),
-    MISSING_PARAMETER("필수 파라미터가 누락되었습니다."),
-    BINDING_FAILED("데이터 바인딩에 실패했습니다.");
+    UNEXPECTED_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "예상치 못한 오류가 발생했습니다. 지원팀에 문의하세요."),
+    INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 오류가 발생했습니다."),
+
+    NULL_ERROR(HttpStatus.BAD_REQUEST, "널 포인터 예외가 발생했습니다."),
+    INVALID_INPUT(HttpStatus.BAD_REQUEST, "유효하지 않은 입력값입니다."),
+    VALIDATION_FAILED(HttpStatus.BAD_REQUEST, "입력값 유효성 검사에 실패했습니다."),
+    MALFORMED_JSON(HttpStatus.BAD_REQUEST, "잘못된 JSON 형식입니다."),
+    MISSING_PARAMETER(HttpStatus.BAD_REQUEST, "필수 파라미터가 누락되었습니다."),
+    BINDING_FAILED(HttpStatus.BAD_REQUEST, "데이터 바인딩에 실패했습니다.");
```
추가적으로 HttpStatus 관리 좋네요 👍

liquidjoo reviewed (approved):
안녕하세요 현준님
피드백 반영 잘해주셨어요 👍 
저와 함께 했던 리뷰들이 도움이 되었길 빌며 마무리하겠습니다!
수고하셨습니다 🙇
